Make getters work with modules

rhenkin opened the pull request:
This addresses #203 by including an optional module id argument to the functions. My understanding is that the communication with JS/Vega side needs the namespace with the module id, but the session$input[[inputId]] reactive does not. Also, unless I got this very wrong, the lines about the inputId_proposed and get_unique_inputId function are redundant because the proposed inputId will always be the first in the vector and, if there are others with the same name, they will be the ones modified, as they will come after inputId. E.g. make.unique(c("this_id", c("this_id","another_id")), sep="_") will result in "this_id", "this_id_1" and "another_id".

ijlyttle commented on R/shiny-getters.R:
```diff
@@ -69,6 +69,7 @@
 #' @param body_value `character` or `JS_EVAL`, the **body** of a JavaScript
 #'   function that Vega will use to handle the signal or event; this function
 #'   must return a value
+#' @param id `character`, optional module id
```
I would call the new argument `moduleId` - I know the [modules documentation](https://shiny.rstudio.com/articles/modules.html) uses `id`, but in the absence of a compelling reason otherwise, I prefer to be more descriptive.

rhenkin replied:
Sure, I don't like id either :)

ijlyttle commented on R/shiny-getters.R:
```diff
@@ -79,31 +80,29 @@
 #'
-vw_shiny_get_signal <- function(outputId, name, body_value = "value") {
+vw_shiny_get_signal <-
+  function(outputId, name, body_value = "value", id = NULL) {
```
If you are going over 80 spaces, our convention is to put the overflowing arguments on the following lines, e.g.:

```r
vw_shiny_get_signal <- function(outputId, name, body_value = "value", 
                                moduleId = NULL) {
}
```

ijlyttle commented on R/shiny-getters.R:
```diff
@@ -79,31 +80,29 @@
   session <- shiny::getDefaultReactiveDomain()
-
-  inputId <- ""
+  ns <- shiny::NS(id)
```
I know this is a picky one, but I would put the `shiny::NS()` call first, before the `shiny::getDefaultReactiveDomain()` call.

ijlyttle commented on R/shiny-getters.R:
```diff
@@ -79,31 +80,29 @@
-
-  inputId <- ""
+  ns <- shiny::NS(id)
+  inputId <- glue::glue("{outputId}_signal_{name}")
```
I think we should not change the `inputId` logic. If I remember right, this is to avoid namespace collisions if you have more than one call getting the same signal, data, event, etc.

rhenkin replied:
Sorry I thought I left a comment about it in this commit, but clearly not. I think the issue I found is that get_unique_inputID will always return inputId unmodified, because make.unique will modify all repeated names except the first. So I think if the aim is to avoid duplicated inputIds we should change the get_unique_inputId function then?

ijlyttle reviewed (changes requested):
Hi @rhenkin,

Thanks for the PR and sorry for the delay. 

I have marked up `vw_shiny_get_signal()`, but you can extrapolate this feedback to other functions.

Happy to discuss further, thanks!

ijlyttle commented:
Just a couple more things :)

On your branch, could you run `usethis::pr_merge_main()`? This will bring in the latest changes I have made, updating Vega-Lite, etc.

As well, could you add a bullet item to `NEWS.md`?

Thanks!

ijlyttle commented:
Hi again @rhenkin, 

I had a little time away-from-keyboard and had a thought.

`inputId` is private and guaranteed to be unique within a `session`.

If we are concerned only with `outputId`, could the namespace function be applied in the function-call? 

For example:

```r
  # inside a module function
  ns <- shiny::NS("my_module")

  rct_value <- vegawidget::vw_shiny_get_signal(ns("my_chart"), "my_signal")
```

Could this work in your case, and if so, perhaps this becomes a matter of documentation rather than modifying functions?

rhenkin commented:
> Hi again @rhenkin,
> 
> I had a little time away-from-keyboard and had a thought.
> 
> `inputId` is private and guaranteed to be unique within a `session`.
> 
> If we are concerned only with `outputId`, could the namespace function be applied in the function-call?
> 
> For example:
> 
> ```r
>   # inside a module function
>   ns <- shiny::NS(id)
> 
>   rct_value <- vegawidget::vw_shiny_get_signal(ns("my_chart"), "my_signal")
> ```
> 
> Could this work in your case, and if so, perhaps this becomes a matter of documentation rather than modifying functions?

I think this was one of the first things I tried, but if you namespace the outputId then inputId will be namespaced, but for some reason in the reactive block at the end of the function you don't want to look for the namespaced inputId in session$input. I also tried modifying inputId and outputId with ns inside the shiny::isolate block and the whole thing only works if ns is used in both.

ijlyttle commented:
Hi @rhenkin - I think there's some stuff that I need to figure out better, for myself, specifically: how does a `session` behave differently in a module from in a "regular" shiny app.

I have no specific action to recommend, just to let you know where I am.

rhenkin commented:
No worries, I don't know either and was only trying to find a workaround to fix the problem, hopefully my comments can help you!

ijlyttle commented:
Hi again @rhenkin - I think I've found a simplification.

The [`session` carries an `ns()` function](https://shiny.rstudio.com/reference/shiny/latest/session.html), so I don't think we need to pass a `moduleId`.

Instead, I think we can do something like this (it works on my computer™️):

```r
vw_shiny_get_signal <- function(outputId, name, body_value = "value") {

  assert_packages("shiny")

  session <- shiny::getDefaultReactiveDomain()
  ns <- session$ns

  inputId <- ""

  # set up an observer to run *once* to add the listener
  shiny::observe({

    shiny::isolate({
      # create unique inputId (set in enclosing environment)
      inputId_proposed <- glue::glue("{outputId}_signal_{name}")
      inputId <<- get_unique_inputId(inputId_proposed, names(session$input))

      # compose_handler_body
      handler_body <-
        vw_handler_signal(body_value) %>%
        vw_handler_add_effect("shiny_input", inputId = ns(inputId)) %>%
        vw_handler_body_compose(n_indent = 0L)

      # add listener
      vw_shiny_msg_addSignalListener(
        ns(outputId),
        name = name,
        handlerBody = handler_body
      )
    })

  })

  # return a reactive that listens to our "private" input
  shiny::reactive({
    session$input[[inputId]]
  })
}
```

This is essentially the solution you propose, but I would like to keep the `get_unique_inputId()` logic to make sure that someone won't step on existing inputs with unfortunate names.

Would you consider this approach?

Thanks!

rhenkin commented:
Hi @ijlyttle
It does seem to work (though I'm a bit confused about why the double module namespacing is needed `session$ns(inputId)` AND works).

Also, I understand you want to avoid the duplicated ID but the function is unlikely to be working as intended. The following code:
`get_unique_inputId(inputId = "module-test_signal_selection", names_input = c("module-test_signal_selection", "otherModule-test_signal_selection2"))`
will return "module-test_signal_selection".
If you are expecting something like "module-test_signal_selection2" then we need to change the get_unique_inputId function.

ijlyttle commented:
Hi @rhenkin,

Using different namespacing methods for the "input" and the output is not ideal. My reasoning is:

- for the output, I understand the convention is for users to call using `ns(outputId)`; I'd like to follow that convention.
- for the "input", this is because the only way (I know of) to communicate information from JS to Shiny is to use `Shiny.setInputValue()` ([ref](https://shiny.rstudio.com/articles/communicating-with-js.html)). As a result, the input is meant to be hidden from the user. But it also needs access to the `ns()` function, so I propose to do it quietly.

As well, this does not change the public API, which makes it easier for us to change in the future (rather than changing it again).

Thanks for the catch that `get_unique_inputId()` does not work the way I had in mind! Of course, I did not test it :/

What do you think of using `uuid::UUIDgenerate()`, instead? Something like this at the top of the functions:

```r
inputId <- uuid::UUIDgenerate()
```

Then, as you have suggested all along, we could get rid of `get_unique_inputId()` and this sort of code inside of `shiny::isolate()`:

```r
# create unique inputId (set in enclosing environment)
inputId_proposed <- glue::glue("{outputId}_signal_{name}")
inputId <<- get_unique_inputId(inputId_proposed, names(session$input))
```
 
We'd also need to "suggest" the uuid package (just as shiny is suggested).

Thanks again for your patience with me on this, I very much appreciate it - especially to let me know that the "unique" isn't doing what I had hoped.

rhenkin commented:
@ijlyttle no worries! happy to help.

I tested removing the double namespacing, but using the session$ns, and it seems to work too (ie. passing the "raw" outputID, then applying the ns inside the isolate block, but not inside the final reactive block). It would look like the following:

`vw_shiny_get_signal <- function(outputId, name, body_value = "value") {

  assert_packages("shiny")

  session <- shiny::getDefaultReactiveDomain()

  inputId <- ""

  # set up an observer to run *once* to add the listener
  shiny::observe({

    shiny::isolate({
      # create unique inputId (set in enclosing environment)
      inputId_proposed <- glue::glue("{outputId}_signal_{name}")
      inputId <<- get_unique_inputId(inputId_proposed, names(session$input))

      # compose_handler_body
      handler_body <-
        vw_handler_signal(body_value) %>%
        vw_handler_add_effect("shiny_input", inputId = session$ns(inputId)) %>%
        vw_handler_body_compose(n_indent = 0L)

      # add listener
      vw_shiny_msg_addSignalListener(
        session$ns(outputId),
        name = name,
        handlerBody = handler_body
      )
    })

  })

  # return a reactive that listens to our "private" input
  shiny::reactive({
    session$input[[inputId]]
  })
}`

I can test this approach in the other get/set functions.

Regarding the unique names, I think the solution may be simpler: get_unique_inputId currently combines input names in the order of: `make_unique(c(proposed_input, old_inputs))`, hence why it does not work (proposed_input will always stay the same). I think if you invert the order to `make_unique(c(old_inputs, proposed_input))` then it should work as intended.

ijlyttle commented:
Hi again @rhenkin,

I like your idea to `make_unique(c(old_inputs, proposed_input))` (I wish I had thought of it), that sounds great!

I think we are almost in agreement on the namespace stuff.

Let me try to rationalize what seems like a weird implementation; let's look at this [example from the Shiny docs](https://shiny.rstudio.com/articles/modules.html#using-renderui-within-modules):

<img width="577" alt="image" src="https://user-images.githubusercontent.com/2800934/195466614-0c3a5453-adf8-49b8-8593-9e20cc00cc36.png">

The convention seems to be that if you call a function from the module, you namespace the input/output, e.g.:

```r
selectInput(ns("col"), "Columns", names(data), multiple = TRUE)
```

I think we should follow that convention, which means that we duplicate the work when we get the `ns()` function for our private input using `shiny::getDefaultReactiveDomain()`. I think we should duplicate the work because our input is a private implementation, which could change in the future.

I think we should not add the namespace here:

```r
vw_shiny_msg_addSignalListener(
  outputId,
  name = name,
  handlerBody = handler_body
)
```

Instead, the user would call:

```r
getSelectedCyl  <- vw_shiny_get_signal(ns("chart"), name = "cyl_sel",  body_value = "value")
```

which would be in line with (what I understand to be the convention).

Does that work?

rhenkin commented:
Hi @ijlyttle I understand your point now. I rolled back a few of the changes, in the end the central fix is adding the session$ns to the inputId. I added a news item and also fixed the unique inputId function.

ijlyttle commented:
Hi again @rhenkin - 

Thanks so much for your work and patience on this. It may seem frustrating that we went through all this for seemingly small changes in the end, I see it as a victory that we sorted through the issues and found the right things to do.

I added your name to NEWS, made a small fix to `get_unique_inputId()` and added the tests I should have added before.

Thanks!
